### ISSUE TITLE
mamba3 utils.py: PTX inline-asm helpers (cos_approx/sin_approx/tanh_approx) with NVPTX 'f' constraints cannot compile on ROCm/AMDGPU

### ISSUE THREAD

**Issue by kryptodogg** [state-spaces/mamba#1018]
## Summary

`mamba_ssm/ops/triton/mamba3/utils.py` implements its fast trig/gating approximations (`cos_approx`, `sin_approx`, `tanh_approx`, `sech2_approx`, `sigmoid_approx`) with **NVIDIA PTX inline assembly** via `tl.inline_asm_elementwise(..., constraints="=f,f")`. The `f` constraint is NVPTX register-class notation; LLVM's AMDGPU backend defines no such class and cannot allocate it. Any Mamba-3 kernel that compiles these helpers therefore **hard-fails on ROCm at first Triton compile** with:

```
error: couldn't allocate output register for constraint 'f'
```

## Environment

- mamba-ssm 2.3.2.post1, source-built (`MAMBA_FORCE_BUILD=TRUE`)
- torch 2.13.0+rocm7.2, Triton 3.7.1
- AMD Radeon RX 6700 XT (gfx1031; `HSA_OVERRIDE_GFX_VERSION=10.3.0`)

## Reproduction (100%, two minimal kernels)

The failure reproduces through the module API (`mamba_ssm.modules.mamba3.Mamba3.forward` → `mamba3_siso_combined`) **and** directly with a kernel that only touches the helpers. Kernel A below fails 100%; kernel B (identical, builtins only) compiles and runs clean on the same box.

**Kernel A — PTX helpers (reproduces the error 100%):**

```python
import torch, triton, triton.language as tl
from mamba_ssm.ops.triton.mamba3.utils import cos_approx, sin_approx

@triton.jit
def k_cos(x_ptr, o_ptr, N: tl.constexpr):
    offs = tl.arange(0, N)
    x = tl.load(x_ptr + offs)
    tl.store(o_ptr + offs, cos_approx(x) + sin_approx(x))

x = torch.randn(64, device="cuda", dtype=torch.float32)
o = torch.zeros(64, device="cuda", dtype=torch.float32)
k_cos[(1,)](x, o, 64)   # -> "couldn't allocate output register for constraint 'f'"
```

**Kernel B — builtins (compiles and runs clean on ROCm):**

```python
import torch, triton, triton.language as tl

@triton.jit
def k_builtin(x_ptr, o_ptr, N: tl.constexpr):
    offs = tl.arange(0, N)
    x = tl.load(x_ptr + offs)
    tl.store(o_ptr + offs, tl.sin(x) + tl.cos(x))

x = torch.randn(64, device="cuda", dtype=torch.float32)
o = torch.zeros(64, device="cuda", dtype=torch.float32)
k_builtin[(1,)](x, o, 64)   # OK
```

## Root cause

`utils.py` lines ~14–52 emit PTX opcodes (`cos.approx.f32 $0, $1;`, `sin.approx.f32 $0, $1;`, `tanh.approx.f32 $0, $1;`) with `constraints="=f,f"`. Per Triton's `inline_asm_elementwise` contract, the assembly string "must match target's assembly format" — NVPTX text is being fed to a backend that must emit AMDGCN, and `f` is not a valid AMDGPU register-constraint letter. This is a source-level construct, not a packaging issue (verified against a `MAMBA_FORCE_BUILD=TRUE` source build), and unrelated to the causal-conv1d source-build fix.

## Fix request

Replace the five helpers with ROCm-portable builtins (`tl.cos`/`tl.sin`; tanh/sech2/sigmoid via `tl.sigmoid` or libdevice), or gate the inline-asm path on the backend (NVIDIA-only) with a builtin fallback for ROCm.

**One numerical-equivalence gate must ship with the fix:** `cos.approx.f32`/`sin.approx.f32` and `tl.cos`/`tl.sin` are *different functions* (PTX fast-approx vs target hardware cos). The swap must be verified **numerically within fp16-accumulation tolerance on both targets (NVIDIA + ROCm)** — not just "compiles" — using kernels A/B above with an fp32 reference, as the fix's own verification artifact.

## Scope

All five helpers in `utils.py` carry the same construct (cos/sin are hit first, in the RoPE; tanh/sech2/sigmoid in gating). The fix should cover the file. Until this lands, the Mamba-3 SISO path cannot compile on ROCm at all.

**Comment by kryptodogg** [comment]
## Follow-up: PTX census + CPU paper-equation reference (both shipped here)

### 1. PTX census — the fix is exactly 4 sites, complete for the numerics path

Grep of the entire installed `mamba_ssm/ops/triton/mamba3/` package for `inline_asm_elementwise`:

```
utils.py:23   cos_approx   -> cos.approx.f32 $0, $1;    ACTIVE
utils.py:43   sin_approx   -> sin.approx.f32 $0, $1;    ACTIVE
utils.py:62   tanh_approx  -> tanh.approx.f32 $0, $1;   ACTIVE
utils.py:81   sech2_approx -> tanh.approx via inline    ACTIVE
utils.py:104  sigmoid_approx -> tanh.approx inline      COMMENTED OUT (uses tl.sigmoid)
```

- **Exactly 4 active PTX sites, all in `utils.py`** — zero inline asm anywhere else in the package (kernel files `mamba3_siso_fwd.py`, `mamba3_siso_bwd.py`, `angle_dt.py`, MIMO kernels, step kernels all use the helpers, never raw asm).
- The original filing overcounted: `sigmoid_approx` and `silu` are **not** PTX — their inline-asm bodies are commented out and they return `tl.sigmoid(x)` (verified in source, and `silu` compiles/runs clean on ROCm).
- So "swap these 4 helpers" (`cos_approx`, `sin_approx`, `tanh_approx`, `sech2_approx`) for `tl.cos`/`tl.sin`/libdevice or a ROCm-guarded builtin fallback is the **complete** fix for the forward numerics path — no deeper bwd/ftz/rcp PTX to discover.

### 2. CPU paper-equation reference — the numerics target a fixer can run against

Since the kernel cannot compile on ROCm, this repo built the paper-equation reference in **pure CPU torch** (no Triton, no PTX-compile boundary): it patches the module's kernel entry point (`mamba3_siso_combined`) with a line-by-line audit of `mamba3_siso_fwd.py`, so the real module plumbing (`in_proj`, `B_norm`/`C_norm` RMSNorm, `ADT`/`DT`/`Trap` trapezoidal discretization, angle cumsum + RoPE, `D`-skip, `Z` SiLU gating, out_proj) runs end-to-end on CPU.

Reference implementation (full script attached in the gist below; core math):

```python
# angle_dt: Angles_Cumsum = cumsum(Angles * DT) mod 2pi
# trapezoidal discretization: gamma = DT * sigmoid(Trap); shifted_gamma = DT[t+1]*(1-sigmoid(Trap[t+1]))
#   k_scaled = RoPE(k) * (shifted_gamma + gamma); qk_dot = sum(Q*K, pre-RoPE) * gamma
# da = ADT * log2(e); da_cs = cumsum(da)
# Form 1 (closed form, one chunk):
#   out[t] = sum_{s<t} (q_t . k_scaled_s) * exp2(da_cs[t]-da_cs[s]) * v_s
#          + (D + qk_dot[t]) * v_t
# Form 2 (recurrent, same state convention as the kernel):
#   m_0 = 0; out[t] = q_t . m_t + (D + qk_dot[t]) * v_t
#   m_{t+1} = exp2(da[t+1]) * (m_t + k_scaled[t] (x) v_t)
```

**Validated self-consistent:** closed-form vs recurrent scan agree to `max_abs = 1.9e-6` in fp32 (pure accumulation-level agreement); module forward is finite, deterministic (byte-equal on repeat), and matches the kernel's state convention exactly (out uses the state *including* decay up to t; the diagonal s=t self-term is masked out and replaced by `(D + qk_dot[t]) * v_t` — a subtlety the naive recurrent scan gets wrong).

**Why this is the right gate for the fix:** a fixer swapping `cos.approx` → `tl.cos` can run this reference on CPU (independent of the Triton path), then compare the fp16 Triton kernel against it on both NVIDIA and ROCm within fp16-accumulation tolerance. Without it, "compiles" is the only verdict available.

Gist with both artifacts: https://gist.github.com/ (see comment thread / repo `chakra/memory.md` 2026-08-16 entries; probes at `/tmp/probe_ptx_cossin.py`, `/tmp/probe_tl_trig.py`, `/tmp/verify_mamba3_siso_reference.py`)

**Comment by kryptodogg** [comment]
## Follow-up: ROCm half of the numerical-equivalence gate — measured profile for `tl.cos`/`tl.sin` (the proposed builtin swap)

The fix request asked that the `cos.approx` → `tl.cos`/`tl.sin` swap be verified numerically within fp16-accumulation tolerance on both targets, with an fp32 reference. On this ROCm box the PTX side cannot compile (that's the boundary this issue is about), so the measurable half is the **builtin path vs an fp64 torch reference** — measured on **gfx1031 (RX 6700 XT), torch 2.13.0+rocm7.2, triton 3.7.1**. This is the ROCm-side target a fixer should run against.

### Measured profile

`tl.cos`/`tl.sin` vs fp64 reference, three angle distributions (RoPE-realistic `cumsum(Angles·DT) mod 2π`, full-range `[-π, π]` N=2^18, near-zero `[0, 0.1]`), fp32 and fp16 inputs:

| Case | fp32 in (max_abs vs fp64) | fp32 in, Triton vs torch fp32 | fp16 in (max_abs vs fp64) | fp16 input-rounding only |
|---|---|---|---|---|
| RoPE-realistic | 2.98e-8 | **0.0 (bitwise)** | 9.0e-5 | 2.4e-5 |
| full-range [-π, π] | 6.89e-8 | **0.0 (bitwise)** | 1.07e-3 | 8.9e-4 |
| near-zero [0, 0.1] | 3.23e-8 | **0.0 (bitwise)** | 2.46e-4 | 3.0e-6 |

(sin tracks cos; the RoPE-realistic sin fp16 max_abs is 1.95e-3, 100% input-rounding — the function adds nothing.)

**Three findings for the fixer:**

1. **The function is bitwise identical to torch fp32 cos/sin on ROCm** — `tl.cos`/`tl.sin` lower to the same device libm; the swap itself introduces **zero** function error at fp32.
2. **fp16 error is entirely quantization, not the trig** — all cases ≤ ~2e-3, well inside the 2e-2 fp16-accumulation tolerance the probe suite uses. **The ROCm side of the gate PASSES with the builtin swap.**
3. **Triton contract: `tl.cos`/`tl.sin` reject fp16 input at compile time** (`Expected dtype ['fp32', 'fp64'] but got fp16`). The swap **must** upcast (`x.to(tl.float32)`) before the trig; the fp16 angle rounding enters at that cast — not in the helper.

### Where this lives

- Repo-owned probe: `chakra/triton_probe/probe_mamba3_trig_numerics.py` (GPU-only, canonical `uv run --env-file .env` invocation; runs the fp64 reference + both input dtypes + the input-rounding isolation). The `tl.cos`/`tl.sin` arms are target-agnostic, so the same probe doubles as the NVIDIA side's builtin arm once the swap lands.
- Companion probes: `chakra/triton_probe/probe_mamba3_siso_boundary.py` (the compile boundary repro from this issue) and `chakra/triton_probe/verify_mamba3_siso_reference.py` (the CPU paper-equation reference from the previous comment, now repo-owned and migrated out of /tmp).

The full combined gate once the swap lands: boundary probe must flip to BOUNDARY-FIXED → run the 4b verdict probe (`chakra/triton_probe/verify_mamba3_siso_4b.py`, FWD max_abs < 2e-2 / rel_l2 < 1e-2, grads < 2e-1) → confirm against the CPU reference (self-consistent 1.9e-6). Local pre-verification of that whole sequence (via a vendored sliver of exactly these 4 helpers) already passed on this box: FWD max_abs=1.364e-03 / rel_l2=8.737e-04 at 4b geometry, all grads ≤ 2.2e-3 rel_l2, deterministic byte-equal — so the numerics verdict is not in doubt when the compile boundary is cleared upstream.

**Comment by kryptodogg** [comment]
Follow-up: the ROCm-side numerical-equivalence gate now covers **all four** active PTX sites in `mamba3/utils.py`. Extending the cos/sin profile posted above with the tanh/sech2 arms, measured on gfx1031 (torch 2.13.0+rocm7.2, triton 3.7.1, AMD Radeon RX 6700 XT) with the same fp64-reference methodology, using the exact synthesis forms the swap proposes (`2*sigmoid(2x)-1`, `1-tanh^2`, both with the defensive `x.to(tl.float32)`).

**`tanh_approx` replacement — `2*sigmoid(2x)-1`** (call site: angle_dt fwd, `tanh(angle) * PI`):

| input | vs fp64 `tanh` max_abs | vs torch fp32 (same input) |
|---|---|---|
| fp32, RoPE-realistic angles | 1.19e-7 | 1.19e-7 |
| fp32, [-4, 4] saturation span | 1.82e-7 | 1.79e-7 |
| fp32, near-zero [0, 0.1] | 1.15e-7 | 1.12e-7 |
| fp16, RoPE angles | 7.16e-6 | (input-rounding only: 3.47e-6) |
| fp16, [-4, 4] | 4.30e-4 | (input-rounding only: 2.03e-4) |

The identity `2*sigma(2x)-1 == tanh(x)` holds to fp32 rounding: the fp32 cross-comparison vs `torch.tanh` on identical input peaks at 1.8e-7 even in the saturation regime, so only the sigmoid path's fp32 rounding separates the synthesis from the reference.

**`sech2_approx` replacement — `1-tanh^2`** (call site: angle_dt bwd, the `dt * PI * sech2(pretanh)` derivative factor):

| input | vs fp64 `1/cosh^2` max_abs | vs torch fp32 (same form) |
|---|---|---|
| fp32, RoPE-realistic angles | 2.38e-7 | 2.39e-7 |
| fp32, [-4, 4] | 3.55e-7 | 3.87e-7 |
| fp32, near-zero [0, 0.1] | 4.94e-8 | 5.96e-8 |
| fp16, RoPE angles | 7.47e-5 | (input-rounding only: 5.68e-8) |
| fp16, [-4, 4] | 4.27e-4 | (input-rounding only: 3.10e-4) |

The `1-tanh^2` form vs the mathematical `1/cosh^2` is a form mismatch, not a numerics problem: invisible at fp64 (< 4e-7), so the PTX's own form can be kept in the swap. The fp16 RoPE-realistic total (7.5e-5) is dominated by the fp16 output-store ulp of sech²≈1, with input-rounding alone at 5.7e-8.

**New Triton contract (affects all four sites):** `tl.sigmoid` **rejects fp16 input at compile time** (`CompilationError`), and the earlier finding that `tl.cos`/`tl.sin` reject fp16 (`Expected dtype ['fp32','fp64'] but got fp16`) stands. The `x.to(tl.float32)` upcast in the proposed swap is therefore **load-bearing for tanh/sech2, not merely defensive** — all four replacement sites must upcast before the builtin, and the fp16 angle/pretanh rounding enters at that cast.

**Gate verdict (2e-2 fp16-accumulation tolerance):** all four replacements pass with a large margin. The worst case across every arm is `tl.sin` fp16 RoPE max_abs = 1.95e-3, and it is fully explained by fp16 input rounding (1.95e-3) — the functions themselves contribute ~0 (cos/sin fp32 cross is bitwise 0.0, tanh/sech2 ≤ 1.8e-7 / 3.9e-7). The full runnable probe (`probe_mamba3_trig_numerics.py`, repo-owned in the DeepASC/Synesthesia tree, canonical ROCm env, gfx1031) reproduces this table; the tanh/sech2 arms cover the saturation span [-4, 4] where the sigmoid synthesis must track tanh.

Combined with the CPU paper-equation reference (comment #issuecomment-5302731436), the numerical-equivalence gate for the swap now has a complete verification target on both the numerics and the compile axes.
